File: lockpick-web/src/app/service/filtro.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Catalogo } from '../model/Catalogo';
import { Observable } from 'rxjs';

@Injectable({
    providedIn: 'root'
})

export class FiltroService{
    constructor(private http: HttpClient){}

    url="http://localhost:8080/api/filtro"

    getCatalogoFiltrato(pMin: number, pMax: number, lastKey: number, batch: number, idCategoria: number, idProduttore: number): Observable<Catalogo[]>{
        if(idCategoria > 0 && idProduttore > 0){
            return this.http.get<Catalogo[]>(this.url + "?idCategoria="+idCategoria+"&idProduttore="+idProduttore+"&pMin="+pMin+"&pMax="+pMax+"&lastKey="+lastKey+"&batch="+batch);
        }else if(idCategoria <= 0 && idProduttore > 0){
            //query senza categoria
            return this.http.get<Catalogo[]>(this.url + "NoCat?idProduttore="+idProduttore+"&pMin="+pMin+"&pMax="+pMax+"&lastKey="+lastKey+"&batch="+batch);
        }else if(idProduttore <= 0 && idCategoria > 0){
            //query senza produttore
            console.log("no Produttore")
            return this.http.get<Catalogo[]>(this.url + "NoProd?pMin="+pMin+"&pMax="+pMax+"&lastKey="+lastKey+"&batch="+batch+"&idCategoria="+idCategoria);
        }else if(idProduttore == 0 && idCategoria == 0){
            //query filtro solo col prezzo
            return this.http.get<Catalogo[]>(this.url + "SoloPrezzo?pMin=" + pMin + "&pMax=" + pMax + "&lastKey=" + lastKey + "&batch=" + batch);
        }
    }

    getProdottiSimiliCategoria(idCategoria: number, idCatalogo: number, lastKey: number, batch: number): Observable<Catalogo[]>{
            return this.http.get<Catalogo[]>(this.url + "SimiliCategoria?idCategoria="+idCategoria+"&idCatalogo="+idCatalogo+"&lastKey="+lastKey+"&batch="+batch);
        }


    getProdottiSimiliProduttore(idProduttore: number, idCatalogo: number, lastKey: number, batch: number): Observable<Catalogo[]>{
        return this.http.get<Catalogo[]>(this.url + "SimiliProduttore?idCategoria="+idProduttore+"&idCatalogo="+idCatalogo+"&lastKey="+lastKey+"&batch="+batch);
    }
}

File: lockpick-rest/src/main/java/com/lockpick/service/FiltroService.java
package com.lockpick.service;

import java.util.List;

import com.lockpick.DTO.DTO;

public interface FiltroService {
	public List<DTO> getProdottiFiltrati(Integer idCategoria,
										 Integer idProduttore,
										 int pMin,
										 int pMax,
										 int lastKey,
										 int batch);
	//senza categoria
	public List<DTO> getProdottiFiltrati(Integer idProduttore,
													 int pMin,
													 int pMax,
													 int lastKey,
													 int batch);
	//senza produttore
	public List<DTO> getProdottiFiltrati(int pMin,
										 int pMax,
										 int lastKey,
										 int batch,
										 Integer idCategoria);
	//solo prezzo
	public List<DTO> getProdottiFiltrati(int pMin, int pMax, int lastKey, int batch);
	
	public List<DTO> getFiltroCategoria(int idCategoria, int idCatalogo, int lastKey, int batch);
	
	public List<DTO> getFiltroProduttore(int idProduttore, int idCatalogo, int lastKey, int batch);
}
